File: source/Population.ts
/*!
 * @author electricessence / https://github.com/electricessence/
 * Licensing: MIT https://github.com/electricessence/Genetic-Algorithm-Platform/blob/master/LICENSE.md
 */

import {Set} from "typescript-dotnet-umd/System/Collections/Set";
import {StringKeyDictionary} from "typescript-dotnet-umd/System/Collections/Dictionaries/StringKeyDictionary";
import {Enumerable} from "typescript-dotnet-umd/System.Linq/Linq";
import {ArgumentNullException} from "typescript-dotnet-umd/System/Exceptions/ArgumentNullException";
import {forEach} from "typescript-dotnet-umd/System/Collections/Enumeration/Enumerator";
import {IEnumerateEach} from "typescript-dotnet-umd/System/Collections/Enumeration/IEnumerateEach";
import {
	Predicate, Action, PredicateWithIndex,
	ActionWithIndex
} from "typescript-dotnet-umd/System/FunctionTypes";
import {IEnumerator} from "typescript-dotnet-umd/System/Collections/Enumeration/IEnumerator";
import {IEnumerableOrArray} from "typescript-dotnet-umd/System/Collections/IEnumerableOrArray";
import {IGenome} from "./IGenome";
import {IPopulation} from "./IPopulation";
import {IGenomeFactory} from "./IGenomeFactory";

export class Population<TGenome extends IGenome>
implements IPopulation<TGenome>, IEnumerateEach<TGenome>
{
	private _population:StringKeyDictionary<TGenome>;

	constructor(private _genomeFactory:IGenomeFactory<TGenome>)
	{
		this._population = new StringKeyDictionary<TGenome>();
	}


	get isReadOnly():boolean
	{
		return false;
	}

	get count():number
	{
		return this._population.count;
	}

	remove(item:TGenome):number
	{
		const p = this._population;
		return item && p.removeByKey(item.hash) ? 1 : 0;
	}

	clear():number
	{
		return this._population.clear();
	}

	contains(item:TGenome):boolean
	{
		const p = this._population;
		return !!item && p.containsKey(item.hash);
	}

	copyTo(array:TGenome[], index?:number):TGenome[]
	{
		if(!array) throw new ArgumentNullException('array');

		// This is a generic implementation that will work for all derived classes.
		// It can be overridden and optimized.
		const e = this._population.getEnumerator();
		while(e.moveNext()) // Disposes when finished.
		{
			array[index++] = e.current!.value;
		}
		return array;
	}

	toArray():TGenome[]
	{
		return this.copyTo([]);
	}

	forEach(action:Action<TGenome>, useCopy?:boolean):number;
	forEach(action:Predicate<TGenome>, useCopy?:boolean):number;
	forEach(action:ActionWithIndex<TGenome>, useCopy?:boolean):number;
	forEach(action:PredicateWithIndex<TGenome>, useCopy?:boolean):number;
	forEach(
		action:ActionWithIndex<TGenome>|PredicateWithIndex<TGenome>,
		useCopy?:boolean):number
	{
		return forEach(useCopy ? this.toArray() : this, action);
	}

	getEnumerator():IEnumerator<TGenome>
	{
		return Enumerable
			.from(this._population)
			.select(o=>o.value)
			.getEnumerator();
	}

	add(potential?:TGenome):void
	{
		if(!potential)
		{
			// Be sure to add randomness in...
			const n = this._genomeFactory.generate();
			if(n) this.add(n);
		}
		else
		{
			let ts:string;
			const p = this._population;
			if(potential && !p.containsKey(ts = potential.hash))
			{
				p.addByKeyValue(ts, potential);
			}
		}
	}

	importEntries(genomes:IEnumerableOrArray<TGenome>):number
	{
		let imported = 0;
		forEach(genomes, o=>
		{
			this.add(o);
			imported++;
		});
		return imported;
	}

	populate(count:number, rankedGenomes?:TGenome[]):void
	{
		//noinspection UnnecessaryLocalVariableJS
		let f = this._genomeFactory;
		// Then add mutations from best in source.
		for(let i = 0; i<count; i++)
		{
			const n = f.generate(rankedGenomes);
			if(n) this.add(n);
		}
	}

	// Provide a mechanism for culling the herd without requiring IProblem to be imported.
	keepOnly(selected:IEnumerableOrArray<TGenome>):void
	{
		const hashed = new Set(
			Enumerable
				.from(selected)
				.select(o => o.hash));

		const p = this._population;
		p.forEach(o=>
		{
			const key = o.key;
			if(!hashed.contains(key))
				p.removeByKey(key);
			
		}, true);
	}
}

export default Population;

File: source/Population.cs
/*!
 * @author electricessence / https://github.com/electricessence/
 * Licensing: MIT https://github.com/electricessence/Genetic-Algorithm-Platform/blob/master/LICENSE.md
 */

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace GeneticAlgorithmPlatform
{
	public class Population<TGenome> : ConcurrentDictionary<string, TGenome>
	where TGenome : IGenome

	{
		private IGenomeFactory<TGenome> _genomeFactory;

		public Population(IGenomeFactory<TGenome> genomeFactory)
		{
			_genomeFactory = genomeFactory;
		}


		public bool IsReadOnly
		{
			get
			{
				return false;
			}
		}


		public void Add(TGenome potential)
		{
			string ts;
			if (potential != null && !ContainsKey(ts = potential.Hash))
			{
				if (!potential.IsReadOnly)
					throw new InvalidOperationException("Adding a genome that is not frozen is prohibited.");

				TryAdd(ts, potential);
			}
		}
		public async Task Add()
		{
			// Be sure to add randomness in...
			Add(await _genomeFactory.GenerateAsync().ConfigureAwait(false));
		}

		public void Add(IEnumerable<TGenome> genomes)
		{
			if (genomes != null)
				foreach (var g in genomes)
					Add(g);
		}

		void AddVariation(IEnumerable<TGenome> rankedGenomes = null)
		{
			if (rankedGenomes != null)
			{
				// Ensure only 1 enumerator created to peek into collection.
				using (var e = rankedGenomes.GetEnumerator())
				{
					if (e.MoveNext())
					{
						var top = e.Current;
						var v = top.NextVariation();
						if (v != null)
							Add((TGenome)v);
					}
				}
			}
		}

		public void Populate(int count, IEnumerable<TGenome> rankedGenomes = null)
		{
			AddVariation(rankedGenomes);

			// Then add mutations from best in source.
			for (var i = 0; i < count; i++)
			{
				Add(_genomeFactory.Generate(rankedGenomes));
			}

		}

		public async Task PopulateAsync(int count, IEnumerable<TGenome> rankedGenomes = null)
		{
			AddVariation(rankedGenomes);

			// Then add mutations from best in source.
			for (var i = 0; i < count; i++)
			{
				Add(await _genomeFactory.GenerateAsync(rankedGenomes));
			}

		}

		// Provide a mechanism for culling the herd without requiring IProblem to be imported.
		public void KeepOnly(IEnumerable<TGenome> selected)
		{
			var hashed = new HashSet<string>(selected.Select(o => o.Hash));
			foreach (var o in this.ToArray()) // Make a copy first...
			{
				TGenome g;
				var key = o.Key;
				if (!hashed.Contains(key))
					TryRemove(key, out g);
			}
		}

	}
}